logr.go

```diff
@@ -395,51 +395,50 @@ func (l Logger) WithContext(ctx context.Context) Logger {
 	return l
 }
 
-// WithContextValues extends the list of context keys and the name for them. When
-// given a context through WithContext later, the Logger will extract the
-// values for these keys and log them as additional key/value pairs
+// WithContextValues extends the list of context handlers. When
+// given a context through WithContext later, the Logger will invoke
+// each handler and log the key/value pairs that they return as
 // as if they had been passed to WithValues.
-func (l Logger) WithContextValues(keys ...ContextKey) Logger {
+func (l Logger) WithContextValues(handlers ...ContextHandler) Logger {
 	if l.sink == discardLogSinkInstance {
 		return l
 	}
 
-	if l.contextKeys == nil {
+	if l.contextHandlers == nil {
 		// Copy the parameters to avoid surprises when the caller changes the content
 		// of the parameter slice later.
-		cp := make([]ContextKey, len(keys))
-		copy(cp, keys)
-		l.contextKeys = &cp
+		cp := make([]ContextHandler, len(handlers))
+		copy(cp, handlers)
+		l.contextHandlers = &cp
 	} else {
 		// We must create a new slice because the existing one is shared between
 		// Logger instances.
-		cp := make([]ContextKey, len(*l.contextKeys)+len(keys))
-		copy(cp, *l.contextKeys)
-		copy(cp[len(*l.contextKeys):], keys)
-		l.contextKeys = &cp
+		cp := make([]ContextHandler, len(*l.contextHandlers)+len(handlers))
+		copy(cp, *l.contextHandlers)
+		copy(cp[len(*l.contextHandlers):], handlers)
+		l.contextHandlers = &cp
 	}
 	return l
 }
 
 func (l Logger) valuesFromContext() []interface{} {
-	if l.ctx == nil || l.contextKeys == nil {
+	if l.ctx == nil || l.contextHandlers == nil {
 		return nil
 	}
-	keysAndValues := make([]interface{}, 0, 2*len(*l.contextKeys))
-	for _, key := range *l.contextKeys {
-		if value := l.ctx.Value(key.Key); value != nil {
-			keysAndValues = append(keysAndValues, key.Name, value)
+	// We don't really know how large this has to be. Let's
+	// assume that each handler returns at most two key/value
+	// pairs, of which each needs two entries.
+	keysAndValues := make([]interface{}, 0, 4*len(*l.contextHandlers))
+	for _, handler := range *l.contextHandlers {
+		for _, keyAndValue := range handler(l.ctx) {
+			keysAndValues = append(keysAndValues, keyAndValue.Key, keyAndValue.Value)
 		}
 	}
 	return keysAndValues
 }
 
-// ContextKey defines the string that is to be used in a key/value pair when
-// logging the value that is stored in a context for the given key.
-type ContextKey struct {
-	Key  interface{}
-	Name string
-}
+// ContextHandler retrieves certain values from the context.
+type ContextHandler func(ctx context.Context) KeysAndValues
```

Comment:
I think you want to return `[]KeyAndValue` - the `KeysAndValues` (pseudostruct) type implies that it will be rendered as a struct, but it is not.  In fact, I don't see why this is not just `[]interface{}` as the return type.

Reply:
`[]KeyAndValue` makes sense to me. I don't like `[]interface{}` because then we are back at having to handle odd number of entries, cast the keys to string or even deal with non-string keys. We don't need such freedom here.

But on the other hand, if `[]interface{}` then just gets passed through to `WithValues`, then it is the better type after all. I'll check...

Reply:
I switched to `[]interface{}` and removed the `KeysAndValues` type from this PR. We can discuss it separately (no PR yet).